Avoid IntRef and other overheads in `ScalaRuntime.{toArray, toObjectArray}`

mkeskells opened the pull request.

mkeskells committed:
don't create an IntRef in toArray

lrytz commented on src/library/scala/runtime/ScalaRunTime.scala:
```diff
@@ -130,13 +130,20 @@ object ScalaRunTime {
+        }
+      }
+      xs foreach copier
+      copier.arr
```
Using an iterator seems more natural - is there a reason not to?

som-snytt replied:
That does seem unnatural. But I'm reminded there is still no lint for `Ref` creation.

mkeskells replied:
I just wrote it is a for loop, because the original was like that. I am not sure where this is implicitly called from. I have changed it to an iterator

Regarding the Ref creation - there are 152 matches for `var .*\n[\W]*for` regex in the codebase. Some of these are in comments and false matches, but there are lots of patterns that it doesn't pick up
Most of the cases that I see in library and compiler are ref creation. 

We just did this case because we see a few 00Gb of IntRef, BooleanRef and ObjectRef and trying to squash them

hrhino replied:
@som-snytt I'm reminded people keep asking for one. #8691

hrhino reviewed (approved):
What's a few hundred gigs between friends?

dwijnand commented:
Isn't storage keep?  Oh, this is memory.  Then I guess: are you saying 640 K wasn't enough?

som-snytt commented:
I changed my previous vote from heart to rocket.

hrhino commented:
👀 as in roll.

retronym commented:
2.13.x appears to use an iterator and while loop already in `Iterator.copyToArray` which underlies `toArray`, so marking this a `[nomerge]` to note this is 2.12.x only.

retronym commented:
> 👀 as in roll.

🙄 as in, "that's just the way 🙄 "

retronym commented on src/library/scala/runtime/ScalaRunTime.scala:
```diff
@@ -130,13 +130,17 @@ object ScalaRunTime {
+        arr(i) = it.next().asInstanceOf[AnyRef]
+        i += 1
+      }
+      arr
```
Let's also refactor `toObjectArray` (above) to the same structure.

som-snytt commented:
I had a theory that retronym never does emoji reactions, perhaps in an effort at impartiality.

Is this the exception proving the rule? (That sounds like a Shapeless pun.)

lrytz commented:
@mkeskells I couldn't find any call to the `toArray` method that you're optimizing, or any place in the compiler where we generate a call to this method. How did you observe it?

We do [emit calls to `toObjectArray`](https://github.com/scala/scala/blob/603a70296daf5b82f81343b4cdf4315c2595eb67/src/compiler/scala/tools/nsc/ast/TreeGen.scala#L175) (which should probably be optimzied in the same way, as suggested by Jason).

retronym committed:
Use a while loop in runtime support for varargs array conversions

retronym committed:
Avoid generic array access in loop in ScalaRuntime

Instead, extends the type-case on the src array to separately
deal with each type of primitive array, which can call into
a specialized variant of the loop.

I've also added a special case for empty arrays.

This compiles as:

```
  // access flags 0x1A
  private final static copy$mCc$sp$1([C)[Ljava/lang/Object;
    // parameter final  src
    ALOAD 0
    ARRAYLENGTH
    ISTORE 1
    ILOAD 1
    ICONST_0
    IF_ICMPNE L0
    GETSTATIC scala/Array$.MODULE$ : Lscala/Array$;
    INVOKEVIRTUAL scala/Array$.emptyObjectArray ()[Ljava/lang/Object;
    ARETURN
   L0
    ILOAD 1
    ANEWARRAY java/lang/Object
    ASTORE 2
    ICONST_0
    ISTORE 3
   L1
    ILOAD 3
    ILOAD 1
    IF_ICMPGE L2
    ALOAD 2
    ILOAD 3
    ALOAD 0
    ILOAD 3
    CALOAD
    INVOKESTATIC scala/runtime/BoxesRunTime.boxToCharacter (C)Ljava/lang/Character;
    AASTORE
    ILOAD 3
    ICONST_1
    IADD
    ISTORE 3
    GOTO L1
   L2
    ALOAD 2
    ARETURN
    MAXSTACK = 4
    MAXLOCALS = 4
```

retronym commented:
I've pushed a change to `toObjectArray` to signficantly tighten up the code we generate.

retronym commented:
@lrytz Could you please review my additional commits?

lrytz commented:
They LGTM. I still wonder about how the `toArray` method is actually used (https://github.com/scala/scala/pull/8688#issuecomment-585662005) - @mkeskells do you know?
